examples/attestation/main.go

```diff
@@ -0,0 +1,308 @@
+	// you can find a high-quality, general-purpose solver for the dns-01
+	// challenge in CertMagic:
+	// https://pkg.go.dev/github.com/caddyserver/certmagic#DNS01Solver
+	client := acmez.Client{
+		Client: &acme.Client{
+			Directory: "https://localhost:9000/acme/attestation/directory",
+			HTTPClient: &http.Client{
+				Transport: &http.Transport{
+					TLSClientConfig: &tls.Config{
+						InsecureSkipVerify: true, // REMOVE THIS FOR PRODUCTION USE!
+					},
+				},
+			},
+			Logger: logger,
+		},
+		ChallengeSolvers: map[string]acmez.Solver{
+			acme.ChallengeTypeDeviceAttest01: attSolver{account},
+		},
+	}
+
+	// If the account is new, we need to create it; only do this once!
+	// then be sure to securely store the account key and metadata so
+	// you can reuse it later!
+	account, err = client.NewAccount(ctx, account)
+	if err != nil {
+		return fmt.Errorf("new account: %v", err)
+	}
+
+	// Do the ACME dance with the created account and get the certificates.
+	certs, err := client.ObtainCertificateUsingCSR(ctx, account, csr)
+	if err != nil {
+		return fmt.Errorf("obtaining certificate: %v", err)
+	}
+
+	// ACME servers should usually give you the entire certificate chain
+	// in PEM format, and sometimes even alternate chains! It's up to you
+	// which one(s) to store and use, but whatever you do, be sure to
+	// store the certificate and key somewhere safe and secure, i.e. don't
+	// lose them!
+	for _, cert := range certs {
+		fmt.Printf("Certificate %q:\n%s\n\n", cert.URL, cert.ChainPEM)
+	}
+
+	return nil
+}
+
+// attSolver is a acmez.Solver That requires you to provide the attestation
+// statement.
+type attSolver struct {
+	account acme.Account
+}
+
+func (s attSolver) Present(ctx context.Context, chal acme.Challenge) error {
+	log.Printf("[DEBUG] present: %#v", chal)
+	return nil
+}
+
+func (s attSolver) CleanUp(ctx context.Context, chal acme.Challenge) error {
+	log.Printf("[DEBUG] cleanup: %#v", chal)
+	return nil
+}
+
+func (s attSolver) Payload(ctx context.Context, chal acme.Challenge) (interface{}, error) {
+	log.Printf("[DEBUG] payload: %#v", chal)
+
+	// Calculate key authorization. This is the data that we need to sign to
+	// validate the device attestation challenge.
+	thumbprint, err := jwkThumbprint(s.account.PrivateKey.Public())
+	if err != nil {
+		return nil, err
+	}
+	keyAuthorization := fmt.Sprintf("%s.%s", chal.Token, thumbprint)
+
+	fmt.Println()
+	fmt.Println("Now you need to sign following keyAuthorization:")
+	fmt.Println(keyAuthorization)
+	fmt.Println()
+	fmt.Println("To do this you can use step-kms-plugin running:")
+	fmt.Printf("echo -n %s | step kms attest --format step \"yubikey:slot-id=82?pin-value=123456\"\n", keyAuthorization)
+	fmt.Println()
+	fmt.Println("Please enter the base64 output and press Enter:")
+	reader := bufio.NewReaderSize(os.Stdin, 4096)
+	attObj, err := reader.ReadString('\n')
+	if err != nil {
+		return nil, err
+	}
+
+	return map[string]string{
+		"attObj": attObj,
+	}, nil
```

Comment:
Am I correct to state this cannot be done with `step-kms-plugin` yet? I don't see the code that signs the echoed `keyAuthorization` and transforms it into the CBOR bytes required in the `attObj`.

Reply:
I have some code that does it, but I'm not convinced yet. Are the `--format step` flag good enough, should I use a different command?

Reply:
Commented on that here: https://github.com/smallstep/step-kms-plugin/pull/36#issuecomment-1419668764. I don't have a better suggestion right now. Maybe I will after testing it with a YubiKey and/or a TPM.